README.md
# minimal-jwebsocket
Minimal Java library for communication between server and web browser, designed as a drop-in replacement for the `java.net` `ServerSocket` and `Socket` classes

## Features
- Drop-in replacement for `java.net` `ServerSocket` and `Socket` classes, able to interact with regular `java.net` `Socket` clients and web browsers
- Automatic HTTP Upgrade handling to WebSocket protocol upon web browser handshake
- Automatic encoding and decoding of traffic between server and web browser client after WebSocket handshaking with the help of background processes

## TODO
- Implement various [Opcodes](https://datatracker.ietf.org/doc/html/rfc6455#section-5.2) for frames other than text
- Implement sending data in individual frames instead of all at once
- Encode traffic from server to client (only encoding from client to server is mandatory)

## Installation
As of version `1.0-3` the `minimal-jwebsocket` library is hosted in the Maven Central OSSRH repository. Installation can be done by including the following dependency in your `pom.xml` file:

```xml
<dependency>
    <groupId>com.antonowycz</groupId>
    <artifactId>minimal-jwebsocket</artifactId>
    <version>1.0-3</version>
</dependency>
```

When building your application, you can replace the `java.net.ServerSocket` class with the `com.antonowycz.websocket.ServerSocket` class and the `java.net.Socket` class with the `com.antonowycz.websocket.Socket` class:

```java
import com.antonowycz.websocket.ServerSocket;   // Extension of java.net.ServerSocket
import com.antonowycz.websocket.Socket;         // Extension of java.net.Socket
```

## Usage

### Server
Using the library to communicate with web browsers is designed to be as straight-forward as interacting normal `java.net` `Socket` clients. Below is a simple example server allowing only once consecutive client to connect at a time, to handle more than one client at a time implement a client handler class with a thread for every client.

```java
...
public class TestServer implements Runnable {
    ...
    @Override
    public void run() {
        while (!socket.isClosed()) {
            // Accept a client
            try (Socket client = socket.accept();
                var br = new BufferedReader(new InputStreamReader(client.getInputStream()));
                var pw = new PrintWriter(client.getOutputStream(), true);
                var sc = new Scanner(br)) {
                // After creating a buffered reader, scanner and printerwriter for the client
                while (!socket.isClosed()) {
                    String input;
                    if ((input = sc.next()) == null) break; // Read input from scanner
                    switch (input) {
                        case "ping":
                            pw.println("pong"); // Write output to printwriter
                            break;
                        default:
                            System.out.println("Received: " + input);
                    }
                }
            } catch (NoSuchElementException e) {
                System.out.println("Client disconnected"); // Scanner is empty -- client disconnected
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }
    ...
}
```

### Client
The code above works both when the client on the receiving end is a regular `java.net.Socket` class, or a web browser with the [RFC 6455](https://datatracker.ietf.org/doc/html/rfc6455) protocol implemented, e.g. Chrome 16+, Firefox 11+, IE 10+.

#### Java
The following unit test from `src/test/java/ServerTest.class` ensures backwards compatibility between the extended `websocket.Socket` class and the `java.net.Socket` class:

```java
@Test
public void backwardsCompatibleResponseTest() {
    try (var client = new java.net.Socket(InetAddress.getLocalHost(), server.getPort());
            var br = new BufferedReader(new InputStreamReader(client.getInputStream()));
            var pw = new PrintWriter(new OutputStreamWriter(client.getOutputStream()), true)) {
        pw.println("ping");
        assertEquals("pong", br.readLine());
        pw.println("a".repeat(126));
        assertEquals("b".repeat(126), br.readLine());
        pw.println("a".repeat(65536));
        assertEquals("b".repeat(65536), br.readLine());
    } catch (IOException e) {
        throw new RuntimeException(e);
    }
}
```

#### Web Browser
The web browser communication is tested through the Selenium UI testing toolkit, using the following web page:

```html
<p id="response"></p>
<button class="websocket-communication disabled" id="ping" onclick="ws.send('ping');">ping</button>
<button class="websocket-communication disabled" id="long2b" onclick="ws.send('a'.repeat(126));">long2b</button>
<button class="websocket-communication disabled" id="long8b" onclick="ws.send('a'.repeat(65536));">long8b</button>
<script>
    function displayOutput(output) {
        document.getElementById('response').innerHTML = output;
    }
    const ws = new WebSocket('ws://localhost:8080');
    ws.onmessage = function(message) {
        displayOutput(message.data);
        switch (message.data) {
            case 'ping': ws.send('pong'); break;
        }
    }
    ws.onopen = function() {
        Array.from(document.querySelectorAll('.websocket-communication.disabled')).forEach((el) => el.classList.remove('disabled'));
    }
</script>
```

The following unit test from `src/test/ServerTest.class` ensures the handshake and bidirectional communication between the web server and server are working correctly:

```java
@Test
public void websocketResponseTest() throws InterruptedException {
    Thread.sleep(2000);
    FirefoxOptions options = new FirefoxOptions();
    options.addArguments("--headless");
    WebDriver driver = new FirefoxDriver(options);
    driver.get(WS_HTML_CLIENT);
    WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
    wait.until(webDriver -> ((JavascriptExecutor) webDriver).executeScript("return document.readyState").equals("complete"));
    wait.until(ExpectedConditions.elementToBeClickable(By.id("ping"))).click();
    wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "pong"));
    driver.findElement(By.id("long2b")).click();
    wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "b".repeat(126)));
    driver.findElement(By.id("long8b")).click();
    wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "b".repeat(65536)));
    driver.quit();
}
```

CHANGELOG.md
# Changelog

## [1.0-3] - 2022-07-31

Revision to initial release.

### Changes
- Attach JavaDoc and sources to release JAR
- Publish to Apache Maven Central Repository

### Features
- Drop-in replacement for `java.net` `ServerSocket` and `Socket` classes, able to interact with regular `java.net` `Socket` clients and web browsers
- Automatic HTTP Upgrade handling to WebSocket protocol upon web browser handshake
- Automatic encoding and decoding of traffic between server and web browser client after WebSocket handshaking with the help of background processes

src/test/java/ServerTest.java
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.io.*;
import java.net.InetAddress;
import java.nio.file.Paths;
import java.time.Duration;

import static org.junit.jupiter.api.Assertions.*;

public class ServerTest {
    private static final int SERVER_PORT = 8080;
    private TestServer server;
    private static final String WS_HTML_CLIENT = Paths.get("src", "test", "web", "TestWSClient.html").toUri().toString();

    @BeforeEach
    public void init() {
        server = new TestServer();
        server.connect(SERVER_PORT);
        server.start();
    }

    @AfterEach
    public void stop() throws IOException {
        server.stop();
    }

    /*
     * Test basic communication between the server and a mock client through the standard java.net.Socket class
     */
    @Test
    public void backwardsCompatibleResponseTest() {
        try (var client = new java.net.Socket(InetAddress.getLocalHost(), server.getPort());
             var br = new BufferedReader(new InputStreamReader(client.getInputStream()));
             var pw = new PrintWriter(new OutputStreamWriter(client.getOutputStream()), true)) {
            pw.println("ping");
            assertEquals("pong", br.readLine());
            pw.println("a".repeat(126));
            assertEquals("b".repeat(126), br.readLine());
            pw.println("a".repeat(65536));
            assertEquals("b".repeat(65536), br.readLine());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    /*
     * Test communication between the server and web browser client, passing this test implies proper HTTP upgrade to websocket protocol
     */
    @Test
    public void websocketResponseTest() throws InterruptedException {
        Thread.sleep(2000);
        FirefoxOptions options = new FirefoxOptions();
        options.addArguments("--headless");
        WebDriver driver = new FirefoxDriver(options);
        driver.get(WS_HTML_CLIENT);
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        wait.until(webDriver -> ((JavascriptExecutor) webDriver).executeScript("return document.readyState").equals("complete"));
        wait.until(ExpectedConditions.elementToBeClickable(By.id("ping"))).click();
        wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "pong"));
        driver.findElement(By.id("long2b")).click();
        wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "b".repeat(126)));
        driver.findElement(By.id("long8b")).click();
        wait.until(ExpectedConditions.textToBePresentInElementLocated(By.id("response"), "b".repeat(65536)));
        driver.quit();
    }
}


src/test/java/TestServer.java
import websocket.ServerSocket;
import websocket.Socket;

import java.io.*;
import java.util.NoSuchElementException;
import java.util.Scanner;

public class TestServer implements Runnable {
    private static final int SERVER_PORT = 8080;
    private ServerSocket socket;

    public static void main(String[] args) {
        TestServer server = new TestServer();
        server.connect(SERVER_PORT);
        server.start();
    }

    public void connect(int port) {
        try {
            this.socket = new ServerSocket(port);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void start() {
        new Thread(this).start();
    }

    public void stop() throws IOException {
        socket.close();
    }

    public int getPort() {
        return socket.getLocalPort();
    }

    @Override
    public void run() {
        try (Socket client = socket.accept();
             var br = new BufferedReader(new InputStreamReader(client.getInputStream()));
             var pw = new PrintWriter(client.getOutputStream(), true);
             var sc = new Scanner(br)) {
            String input;
            while (!socket.isClosed()) {
                if ((input = sc.next()) == null) break;
                switch (input) {
                    case "ping":
                        System.out.println("Received ping request -- sending pong response");
                        pw.println("pong");
                        break;
                    default:
                        switch (input.length()) {
                            case 126:
                                System.out.println("Received long request with 2 length bits -- sending back response with 2 length bits");
                                pw.println("b".repeat(126));
                                break;
                            case 65536:
                                System.out.println("Received long request with 8 length bits -- sending back response with 8 length bits");
                                pw.println("b".repeat(65536));
                                break;
                            default:
                                System.out.println(input);
                        }
                }
            }
        } catch (NoSuchElementException e) {
            System.out.println("Client disconnected -- quitting");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}


src/main/java/websocket/Socket.java
package websocket;

import websocket.preprocessors.WSInputStream;
import websocket.preprocessors.WSOutputStream;

import java.io.IOException;
import java.io.OutputStream;
import java.net.SocketImpl;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

public class Socket extends java.net.Socket {
    private boolean isWebSocketConnection;
    private WSInputStream in;
    private WSOutputStream out;
    public Socket(SocketImpl socket) throws IOException {
        super(socket);
    }

    public WSInputStream getInputStream() throws IOException {
        return in;
    }

    public OutputStream getOutputStream() throws IOException {
        return out;
    }

    /*
     * InputStream and OutputStream pre-processors are created in this method instead of the constructor because upon
     * creation of the object a connection has not yet been established, this happens during the ImplAccept() call.
     */
    public void createPreProcessors() throws IOException {
        in = new WSInputStream(this, super.getInputStream()); // Create pre-processor thread for the input stream
        out = new WSOutputStream(this, super.getOutputStream()); // Create pre-processor thread for the output stream
    }

    @Override
    public String toString() {
        return (isWebSocketConnection ? "Web" : "") + super.toString();
    }

    /*
     * Upgrade the current socket to a websocket connection by sending an HTTP/1.1 101 Switching Protocol response.
     * Documentation: https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_a_WebSocket_server_in_Java#handshaking
     */
    public void upgradeWebsocket(String key) throws NoSuchAlgorithmException, IOException {
        // Build response
        byte[] response = ("HTTP/1.1 101 Switching Protocols\r\n"
                + "Connection: Upgrade\r\n"
                + "Upgrade: websocket\r\n"
                + "Sec-WebSocket-Accept: "
                + Base64.getEncoder().encodeToString(
                        MessageDigest.getInstance("SHA-1").digest(
                                (key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").getBytes(StandardCharsets.UTF_8)))
                + "\r\n\r\n").getBytes(StandardCharsets.UTF_8);
        out.write(response); // Send response
        isWebSocketConnection = true; // Treat this connection as upgraded
    }

    /*
     * Returns whether the current socket has been upgraded to a websocket connection,
     * implies the pre-processors should encode and decode the raw streams.
     */
    public boolean isWebSocketConnection() {
        return isWebSocketConnection;
    }
}


src/main/java/websocket/ServerSocket.java
package websocket;

import java.io.IOException;
import java.net.SocketException;
import java.net.SocketImpl;

public class ServerSocket extends java.net.ServerSocket {
    /*
     * The websocket.ServerSocket class of the minimal-jwebsocket packages provides a method to interact with browser clients
     * whilst being backwards compatible with the java.net websocket.ServerSocket implementation.
     */
    public ServerSocket(int port) throws IOException {
        super(port);
    }

    /*
     * Accept a new connection to the current websocket.ServerSocket object.
     * Adapted from java.net.websocket.ServerSocket to use the local websocket.Socket class instead of java.net.websocket.Socket
     */
    @Override
    public Socket accept() throws IOException {
        if (isClosed())
            throw new SocketException("websocket.Socket is closed");
        if (!isBound())
            throw new SocketException("websocket.Socket is not bound yet");
        Socket s = new Socket((SocketImpl) null);
        implAccept(s);
        return s;
    }

    /*
     * Call the java.net.websocket.Socket.implAccept() method to create input and output stream, then create pre-processors for
     * both streams which will handle the encoding and decoding for when the connection is upgraded to type websocket.
     */
    protected final void implAccept(Socket s) throws IOException {
        super.implAccept(s);
        s.createPreProcessors();
    }
}


src/main/java/websocket/preprocessors/WSOutputStream.java
package websocket.preprocessors;

import websocket.Socket;

import java.io.*;
import java.nio.ByteBuffer;
import java.util.NoSuchElementException;
import java.util.Scanner;

public class WSOutputStream extends java.io.OutputStream implements Runnable {
    private final Socket socket;
    private final java.io.OutputStream out;
    private final PipedInputStream pipedIn;
    private final PipedOutputStream pipedOut;

    public WSOutputStream(Socket s, OutputStream out) throws IOException {
        this.socket = s;
        this.out = out;
        this.pipedIn = new PipedInputStream();
        this.pipedOut = new PipedOutputStream(pipedIn);
        new Thread(this).start();
    }

    @Override
    public void write(int b) throws IOException {
        pipedOut.write(b);
    }

    @Override
    public void write(byte[] b) throws IOException {
        pipedOut.write(b);
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        pipedOut.write(b, off, len);
    }

    /*
     * Encoding is done according to the RFC 6455 specification, i.e. the WebSocket Protocol
     * Documentation: https://www.rfc-editor.org/rfc/rfc6455#section-6.1
     * Example: https://stackoverflow.com/questions/8125507/how-can-i-send-and-receive-websocket-messages-on-the-server-side
     */
    private byte[] encodeOutgoingTraffic(String input) {
        int len = input.length();
        int lenBit = len;
        int numLenBits = 0;
        /*
         * If len <= 125 the message length can be encoded in a single byte.
         * The most significant bit (128) implies whether the message is encoded or not,
         * this is not mandatory for server-to-client traffic but is for client-to-server communication.
         */
        if (lenBit > 125) {
            if (lenBit > 65535) {
                lenBit = 127;
                numLenBits = 8; // The following 8 bytes contain the message length
            } else {
                lenBit = 126;
                numLenBits = 2; // The following 2 bytes contain the message length
            }
        }
        byte[] encoded = new byte[2 + numLenBits + input.length()];
        if (lenBit > 125) {
            // Create a set of bytes from an integer and copy said bytes to the encoded byte array
            ByteBuffer bb = ByteBuffer.allocate(numLenBits);
            if (lenBit == 126) bb.putShort((short) len); else bb.putLong(len);
            System.arraycopy(bb.array(), 0, encoded, 2, numLenBits);
        }
        encoded[0] = (byte) 129; // 10000001 for a text frame
        encoded[1] = (byte) lenBit; // 0 - 125 for actual message length, 126 for 2 bytes and 127 for 8 bytes
        // Add the bytes of the actual message to the encoded byte array
        System.arraycopy(input.getBytes(), 0, encoded, 2 + numLenBits, len);
        return encoded;
    }

    @Override
    public void run() {
        try (var sc = new Scanner(pipedIn)) {
            String data;
            // If receiving HTTP/1.1 101 Switching Protocols, pass complete request
            if ((data = sc.nextLine()).startsWith("HTTP/1.1 101 Switching Protocols"))
                out.write((data + '\n' + sc.useDelimiter("\\r\\n\\r\\n").next() + "\r\n\r\n").getBytes());
            else out.write((data + '\n').getBytes());
            while (!socket.isClosed()) out.write(socket.isWebSocketConnection() ? encodeOutgoingTraffic(sc.nextLine()) : (sc.nextLine() + '\n').getBytes());
        } catch (NoSuchElementException | IOException ignored) { /* Client disconnected */ }
    }
}


src/main/java/websocket/preprocessors/WSInputStream.java
package websocket.preprocessors;

import websocket.Socket;

import java.io.*;
import java.nio.ByteBuffer;
import java.security.NoSuchAlgorithmException;
import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WSInputStream extends java.io.InputStream implements Runnable {
    private final Socket socket;
    private final InputStream in;
    private final PipedInputStream pipedIn;
    private final PipedOutputStream pipedOut;
    public WSInputStream(Socket s, InputStream in) throws IOException {
        this.socket = s;
        this.in = in;
        this.pipedIn = new PipedInputStream();
        this.pipedOut = new PipedOutputStream(pipedIn);
        new Thread(this).start();
    }

    @Override
    public int read() throws IOException {
        return pipedIn.read();
    }

    @Override
    public int read(byte[] b) throws IOException {
        return pipedIn.read(b);
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        return pipedIn.read(b, off, len);
    }

    @Override
    public byte[] readAllBytes() throws IOException {
        return pipedIn.readAllBytes();
    }

    @Override
    public byte[] readNBytes(int len) throws IOException {
        return pipedIn.readNBytes(len);
    }

    @Override
    public int readNBytes(byte[] b, int off, int len) throws IOException {
        return pipedIn.readNBytes(b, off, len);
    }

    /*
     * Receive and decode requests sent by client, Mozilla has a well-documented implementation on their website
     * Documentation: https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_a_WebSocket_server_in_Java#decoding_messages
     */
    public String decodeIncomingTraffic() throws IOException {
        byte[] message;
        int head = in.read(); // First byte gives information on the message itself
        int FIN = (head >> 7) & 1; // Get most significant bit, FIN 1 means this is the whole message
        int opCode = head & (0xf); // Get first 4 bits, Opcode 0x1 means this is a text
        int RSV1 = (head >> 6) & 1;
        int RSV2 = (head >> 5) & 1;
        int RSV3 = (head >> 4) & 1;
        if (FIN != 1 || RSV1 != 0 || RSV2 != 0 || RSV3 != 0 || opCode != 1) // TODO: Implement various Opcodes
            throw new IOException("Unsupported message frame " + String.format("%x", head));
        int len = in.read() - 128; // Second byte gives information on the length of the message
        if (len > 125) {
            // If the second byte minus 128 is between 0 and 125, this is the length of the message
            byte[] arr;
            int numLenBits = len == 126 ? 2 : 8;
            // If it is 126, the following 2 bytes (16-bit unsigned integer)
            // If 127, the following 8 bytes (64-bit unsigned integer, the most significant bit MUST be 0)
            arr = new byte[numLenBits];
            for (int i = 0; i < numLenBits; i++) arr[i] = (byte) in.read();
            ByteBuffer bb = ByteBuffer.wrap(arr);
            len = len == 126 ? (int) bb.getShort() : (int) bb.getLong();
        }
        byte[] key = new byte[4];
        for (int i = 0; i < 4; i++) key[i] = (byte) in.read();
        message = new byte[len];
        for (int i = 0; i < len; i++) message[i] = (byte) (in.read() ^ key[i & 0x3]);
        return new String(message);
    }

    @Override
    public void run() {
        try (var sc = new Scanner(in); var pw = new PrintWriter(pipedOut, true)) {
            String data;
            if ((data = sc.nextLine()).startsWith("GET") && data.endsWith("HTTP/1.1")) {
                // Upon receiving an HTTP GET request, attempt to upgrade the connection to type websocket
                data = sc.useDelimiter("\\r\\n\\r\\n").next();
                Matcher match = Pattern.compile("Sec-WebSocket-Key: (.*)").matcher(data);
                if (match.find()) socket.upgradeWebsocket(match.group(1));
            } else pw.println(data);
            while (!socket.isClosed()) pw.println(socket.isWebSocketConnection() ? decodeIncomingTraffic() : sc.nextLine());
        } catch (NoSuchElementException | IOException ignored) { /* Client disconnected */ } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}
